src/sudoku.rs

```diff
@@ -0,0 +1,244 @@
+                    let mut it = (0 .. BOARD_WIDTH)
+                        .filter(|&x| puzzle.map[y][x] & bit != 0);
+                    let next = it.next();
+                    if next.is_none() || it.next().is_some() { continue }
+                    next
+                };
+                puzzle.map[y][next.unwrap()] = bit;
+            }
+
+            // Check each columns
+            for x in (0 .. BOARD_WIDTH) {
+                let next = {
+                    let mut it = (0 .. BOARD_HEIGHT)
+                        .filter(|&y| puzzle.map[y][x] & bit != 0);
+                    let next = it.next();
+                    if next.is_none() || it.next().is_some() { continue }
+                    next
+                };
+                puzzle.map[next.unwrap()][x] = bit;
+            }
+
+            // Check each groups
+            for y0 in iter::range_step(0, BOARD_HEIGHT, GROUP_WIDTH) {
+                for x0 in iter::range_step(0, BOARD_WIDTH, GROUP_HEIGHT) {
+                    let next = {
+                        let mut it = idx_in_grp
+                            .iter()
+                            .map(|&(dx, dy)| (x0 + dx, y0 + dy))
+                            .filter(|&(x, y)| puzzle.map[y][x] & bit != 0);
+                        let next = it.next();
+                        if next.is_none() || it.next().is_some() { continue }
+                        next
+                    };
+                    let (x, y) = next.unwrap();
+                    puzzle.map[y][x] = bit;
+                }
+            }
+        }
+
+        // Loop until no cells can be filled.
+        if puzzle == bkup { break }
+    }
+
+    let it = (0 .. BOARD_HEIGHT * BOARD_WIDTH)
+        .map(|i| (i % BOARD_WIDTH, i / BOARD_WIDTH))
+        .map(|(x, y)| (x, y, puzzle.map[y][x].count_ones() as BITS))
+        .collect::<Vec<_>>();
+
+    // If some cells have no possible number, there is no answer.
+    if it.iter().any(|&(_x, _y, cnt)| cnt == 0) { return vec![]; }
+
+    // If all cells have exact one possible number, this is a answer.
+    if it.iter().all(|&(_x, _y, cnt)| cnt == 1) { return vec![puzzle]; }
+
+    // Find the first undetermined cell.
+    let (x, y, _cnt) = *it.iter()
+        .filter(|& &(_x, _y, cnt)| cnt > 1)
+        .min_by(|& &(_x, _y, cnt)| cnt)
+        .unwrap();
+
+    let mut answers = vec![];
+    for n in (0 .. MAX_NUMBER) {
+        let bit = 1 << n;
+        if puzzle.map[y][x] & bit == 0 { continue }
+
+        // Assuming the number at (x, y) is `n`, try to solve the problem again.
+        // If some answers are found, append them to the `answers`.
+        let mut p2 = puzzle;
+        p2.map[y][x] = bit;
+        answers.extend(solve_sudoku(p2).into_iter());
+    }
+    answers
+}
+
+const INPUT: &'static str = "
+    850002400
+    720000009
+    004000000
+    000107002
+    305000900
+    040000000
+    000080070
+    017000000
+    000036040
+";
+
+fn main() {
```

Comment:
you will need a `#[cfg(not(test))]` on `main` to avoid this warning:

    src/sudoku.rs:215:1: 223:2 warning: function is never used: `main`, #[warn(dead_code)] on by default
    src/sudoku.rs:215 fn main() {